content/news: add GSoC 2022 Work Product (#279)
mixxxdj/website: content/news/2022-09-02-gsoc-final-blog-david-chocholaty.md

of an audio signal. Thanks to the long working period, the project was expanded
with the implementation of the Group Delay handling for the effect chain.

#### Motivation
Before the effect is implemented, the pitch could be changed using the deck's
rate slider separately only. This imposes significant restrictions on the ways
other effects can interact with the sound. In this project, the new Pitch Shift
effect is introduced in the built-in effects, which can be used
in the effect chain. It implements the wish
[Add a Transpose / Pitch Shift effect](https://github.com/mixxxdj/mixxx/issues/7389)
for the Mixxx application. The effect has to work with the effect chain API.
With that, the other extensional options can be used. Primarily, it allows
for the user to use a much wider range, than the Pitch Shifter
for the deck player. Due to the produced latency based on the pitch processing,
the production delay has to be handled for the wet/dry or wet+dry mode
to ensure that the original (dry) and processed (wet) signals overlap.

#### Background
List of terms from the field of music, sound processing
and development in general:

* Scale

>   * In music theory, a scale is any set of musical notes ordered

Reviewer: This indentation is lost in the rendered html

Author: Thank you for the hint. Now, it should be fixed.

> title: GSoC 2022 Work Product- Pitch Shift effect and Group delay handling

Reviewer 2: ```suggestion
title: GSoC 2022 Work Product - Pitch Shift effect and Group delay handling
```

authors: David Chocholatý
status: draft
tags: TODO(davidchocholaty) add tags
comments: yes

> 

Reviewer 2: I think we should add a disclaimer that this blogpost primarily serves as documentation to google and thus goes into a lot more detail our other blogposts.

the mentioned Pitch Shifter has no wide range and is not really useful
for the interesting DJ skills on live sets. In this project, the new Pitch Shift
effect is introduced in the built-in effects, which can be used
in the effect chain. It implements the wish
[Add a Transpose / Pitch Shift effect](https://github.com/mixxxdj/mixxx/issues/7389)
for the Mixxx application. The effect has to work with the effect chain API.
With that, the other extensional options can be used. Primarily, it allows
for the user to use a much wider range, than the Pitch Shifter
for the deck player. Due to the produced latency based on the pitch processing,
the production delay has to be handled for the wet/dry or wet+dry mode
to ensure that the original (dry) and processed (wet) signals overlap.

#### Background
List of terms from the field of music and sound processing:
<ul>
    <li>Scale
        <ul>
            <li>In music theory, a scale is any set of musical notes ordered
            by fundamental frequency or pitch. The scale ordered
            by increasing pitch is an ascending scale, and a scale ordered
            by decreasing pitch is a descending scale.</li>
        </ul> 

>     </li>

Reviewer 2: markdown is simpler and should work just as well.
```suggestion
 * Scale
   * In music theory, a scale is any set of musical notes ordered
    by fundamental frequency or pitch. The scale ordered
    by increasing pitch is an ascending scale, and a scale ordered
    by decreasing pitch is a descending scale.
```

            The audio buffer size is the amount of time allowed for the computer
            to process the audio of your sound card or audio interface.</li>
        </ul>  
    </li>
    <li>Latency
        <ul>
            <li>Latency, from a general point of view, is a time delay
            between the cause and the effect of some physical change
            in the system being observed.</li>
        </ul>  
    </li>
    <li>Minimum requirements policy
        <ul>
            <li>The main branch must build with the packages in the latest
            Ubuntu LTS which has had at least one point release.</li>
        </ul>  
    </li>
</ul> 

#### Pull requests and issues
[#4775](https://github.com/mixxxdj/mixxx/pull/4775)
- PitchShiftEffect: add independent effect

The PR adds an independent effect to Mixxx's built-in effects.
The implementation uses the RubberBand library for changing a pitch
of an input track. The effect works in real-time mode and adheres
to the “push model” implementation. It means that the input data are offered
to the RubberBand library instead of that the library requires the amount
of input data.

> <hr />

Reviewer 2: ```suggestion
---
```

title: GSoC 2022 Work Product - Pitch Shift effect and Group delay handling
authors: David Chocholatý
status: draft
tags: TODO(davidchocholaty) add tags
comments: yes

#### Introduction
The project implements the Pitch Shift effect for the Mixxx DJ
application. The Pitch Shift effect raises or lowers the original pitch
of an audio signal. Thanks to the long working period, the project was expanded
with the implementation of the Group Delay handling for the effect chain.

#### Motivation
Before the effect is implemented, the pitch could be changed using the master
Pitch Shifter for each deck player separately only. Unfortunately,
the mentioned Pitch Shifter has no wide range and is not really useful

> for the interesting DJ skills on live sets. In this project, the new Pitch Shift

Reviewer 2: ```suggestion
Before the effect is implemented, the pitch of a track could be changed using the decks
rate slider separately only. This imposes significant restrictions in the ways other effects can interact with the sound. In this project, the new Pitch Shift
```

Both of these statements is not really true, there are a lot of creative things DJs have done with the rate adjustment of a turntable. Also it can shift by quite a lot if configured so in the settings.

effect is introduced in the built-in effects, which can be used
in the effect chain. It implements the wish
[Add a Transpose / Pitch Shift effect](https://github.com/mixxxdj/mixxx/issues/7389)
for the Mixxx application. The effect has to work with the effect chain API.
With that, the other extensional options can be used. Primarily, it allows
for the user to use a much wider range, than the Pitch Shifter
for the deck player. Due to the produced latency based on the pitch processing,
the production delay has to be handled for the wet/dry or wet+dry mode
to ensure that the original (dry) and processed (wet) signals overlap.

#### Background
List of terms from the field of music, sound processing
and development in general:
<ul>
    <li>Scale
        <ul>
            <li>In music theory, a scale is any set of musical notes ordered
            by fundamental frequency or pitch. The scale ordered
            by increasing pitch is an ascending scale, and a scale ordered
            by decreasing pitch is a descending scale.</li>

>         </ul>

Reviewer 2: prefer the simpler Markdown constructs if possible.
```suggestion
* Scale
  * In music theory, a scale is any set of musical notes ordered
    by fundamental frequency or pitch. The scale ordered
    by increasing pitch is an ascending scale, and a scale ordered
    by decreasing pitch is a descending scale.
```

    </li>
    <li>Octave
        <ul>
            <li>In music, an octave is an interval between one musical pitch
            and another with double its frequency.</li>
        </ul>
    </li>
    <li>Interval
        <ul>

>             <li>An interval in music is the distance in pitch

Reviewer 2: "distance in pitch" doesn't make sense to me. What unit is that supposed to be?

Author: Replaced with a new version: In music theory, an interval is a difference in pitch between two sounds. In western scales, intervals are most commonly differences between whole tones and semitones.

            between any two notes.</li>
        </ul>
    </li>
    <li>Semitone
        <ul>
            <li>A semitone is a distance in pitch between a note
            and the very next note, higher or lower. It is the smallest interval

>             in western music.</li>

Reviewer 2: ```suggestion
            <li>A semitone is a distance in pitch between a note
            and the very next note, higher or lower. It is the smallest interval
            in most western scales.</li>
```

        </ul>  
    </li>
    <li>Chromatic scale
        <ul>
            <li>A chromatic scale is a set of twelve pitches used
            in tonal music, with notes separated by the interval
            of a semitone.</li>
        </ul>
    </li>
    <li>Pitch
        <ul>
            <li>Pitch is a perceptual property of sounds that allows
            their ordering on a frequency-related scale. The pitch
            is the quality that makes it possible to judge sounds as "higher"
            and "lower" in the sense associated with musical melodies.</li>
        </ul>  
    </li>
    <li>Wet and dry signals
        <ul>
            <li>Dry sound signals refer to the raw or unprocessed sounds
            that usually come from a direct recording. On the other hand,
            wet sounds refer to the processed sound/signal.</li>
        </ul>  
    </li>
    <li>Ring buffer (circular buffer)
        <ul>
            <li>In computer science, a circular buffer or ring buffer
            is a data structure that uses a single, fixed-size buffer
            as if it were connected end-to-end. This structure lends itself
            easily to buffering data streams.</li>

>         </ul>  

Reviewer 2: I think we should leave citations for snippets ripped from other sources.

Author: I'm afraid, that I do not fully understand what you mean. Do you mean to remove the description from here and rather use the link in the text on the citation source?

Reviewer 2: Well, It's common practice when you cite something (just as you cited wikipedia here) you usually provide a reference to the source right where you cited it (or provide at least a footnote). At least thats how I learned it in school in germany, there are probably regional differences though.

Here's an example on how footnotes work in pelican (which we use for building the website): https://github.com/getpelican/pelican/blob/master/pelican/tests/content/article_with_markdown_and_footnote.md

Reviewer 2: So Im just requesting that you add the annotations that link the quote to the source. 
```quote
In computer science, a circular buffer or ring buffer
is a data structure that uses a single, fixed-size buffer
as if it were connected end-to-end. This structure lends itself
easily to buffering data streams.[^1]
```
At the page bottom:
```
[^1]: https://en.wikipedia.org/wiki/Circular_buffer
```

Author: Perfect, I just didn't understand, that you have this approach in mind. :+1: Yeah, I know it a lot from technical papers.

Author: > At least thats how I learned it in school in germany, there are probably regional differences though.

Just for interest, we use it in the Czech Republic's universities as well.

    </li>
    <li>Audio buffer
        <ul>
            <li>An audio buffer holds a single buffer of audio data.
            The audio buffer size is the amount of time allowed for the computer

>             to process the audio of your sound card or audio interface.</li>

Reviewer 2: ```suggestion
            <li>An audio buffer holds a fixed size amount of sampled audio data.
            The audio buffer size determines the time allowed for the computer
            to process the audio data. Thus it also determines Latency.</li>
```

        </ul>  
    </li>
    <li>Latency
        <ul>
            <li>Latency, from a general point of view, is a time delay
            between the cause and the effect of some physical change

>             in the system being observed.</li>

Reviewer 2: IMO too general of an explanation. We should either replace it by an audio specific explanation or at least add a concrete example.

Author: Replaced with a new version:

In the audio world, “latency” is another word for “delay”.
     The latency (time) of an audio system refers to the time difference
     from the moment a signal is fed into the system, to the moment it appears
     at the output. For example, audio latency is when there’s a noticeable
     delay between the sound being played and the moment it reaches
     the speakers. Depending on the application, such a delay can have various
     effects. Usually, the aim is to achieve the lowest possible latency.

        </ul>  
    </li>
    <li>Minimum requirements policy
        <ul>
            <li>The main branch must build with the packages in the latest
            Ubuntu LTS which has had at least one point release.</li>
        </ul>  

>     </li>

Reviewer 2: not important for understanding the article IMO. We should just link to the github wiki page when the term is mentioned in the document body.

The latency has to be handled for the wet/dry and wet+dry modes that the dry
and wet signals overlapped. The structure for delay reporting from the effects
into the effect chain was implemented. With that, the dry signal delaying
to overlap with the wet signal was implemented too. Because it is
a critical part of the application engine performance, the tests and benchmarks
were included in the development.

In this PR, the std::span was newly introduced into the Mixxx app code
with the design proposal and cooperation of my Mentor. The util for working
with spans was implemented, so other developers can easily work with spans
directly from the custom Mixxx data structures. With that, the Mixxx code
is being upgraded using the C++20 standard.

<hr />

[#4848](https://github.com/mixxxdj/mixxx/pull/4848)
- Fix EngineDelay and EngineFilterDelay modulo calculation documentation

Based on the code changes in the EngineEffectsDelay and discussion
with my mentor, the explanation commentary was added to two other Mixxx
structures working on a quite similar principle.

<hr />

[#4852](https://github.com/mixxxdj/mixxx/pull/4852)

> - RingDelayBuffer: ring buffer for delay handling

Reviewer 2: since these PRs don't belong to the same repo all the time, I'd prepend the repo name.
```suggestion
[mixxx#4852](https://github.com/mixxxdj/mixxx/pull/4852)
- RingDelayBuffer: ring buffer for delay handling
```

<hr />

[#10840](https://github.com/mixxxdj/mixxx/pull/10840)
- EngineEffectsDelay: introduce ring delay buffer

The implemented optimized ring buffer data structure for delay handling is built
into the effect chain handling structure. With the use of the new
data structure, the delay handling performance is highly improved based on
the benchmark measurements. Unfortunately, the PR was not merged during
the coding period due to a failing test for the macOS CI (based on the inner
rounding problem for zero value). At the same time, Mixxx's macOS CI
started crashing during the configuration stage because of an issue
that the workflow runner has changed. For that reason, the bug fix
couldn't be tested and the PR was not merged in time.

<hr />

[#10843](https://github.com/mixxxdj/mixxx/pull/10843)
- RingDelayBufferTest: refactor includes and span creation

The tests for the RingDelayBuffer are refactored and the span creations
are deduplicated.

<hr />

[#10858](https://github.com/mixxxdj/mixxx/pull/10858)
- PitchShiftEffect: add description comments

Added the comments for the Pitch shift effect processing.

<hr />

> 

Reviewer 2: I'd also add a mention of the PR to this blog post here since its technically also work done for mixxx imo.

<ul>
	<li>Pitch knob</li>
	<li>Range knob</li>
	<li>Semitones mode</li>
	<li>Formant preserving</li>
</ul>

The Pitch knob changes the Pitch of the track up or down. For the default
middle position, the track pitch is unchanged. The Range knob ensures setting
the range of the Pitch knob. The Pitch knob based on the range setting can work
from zero range to 2 octaves range. These two knobs work similarly to the real
professional Pioneer DJM-900NX2 mixer which is widely used in clubs for live
DJ mixing. Then, the Semitones mode toggle was added. This option sets the scale
of the Pitch knob. The knob can work in two modes: continuous or semitones mode.
In the semitones mode, the pitch is changed based on the chromatic scale
in a musical way. Otherwise, the pitch is processed continuously working,
which is the default approach of the used RubberBand library. By default,
the semitones mode is on. As last, the Formant preserving option was added
which uses the RubberBand API same called option. It preserves
the resonant frequencies (formants) of the human vocal tract
and other instruments (compensates for “chipmunk” or “growling” voices).

Related files:

> - [pitchshifteffect.cpp](https://github.com/mixxxdj/mixxx/blob/main/src/effects/backends/builtin/pitchshifteffect.cpp)

Reviewer 2: we should probably use the permalink instead:
```suggestion
- [pitchshifteffect.cpp](https://github.com/mixxxdj/mixxx/blob/d232119d8fd82e9bdaaf2958e847593156b261f0/src/effects/backends/builtin/pitchshifteffect.cpp)
```

- [pitchshifteffect.h](https://github.com/mixxxdj/mixxx/blob/main/src/effects/backends/builtin/pitchshifteffect.h)

Screenshots:

*Pitch shift effect in the Mixxx app*
![Pitch shift effect in the Mixxx app]({static}/images/news/mixxx-pitch-shift-app.png)

*Pitch shift effect in the effect chain*
![Pitch shift effect in the effect chain]({static}/images/news/mixxx-pitch-shift-effect.png)

##### *Group delay handling*

As a project extension, the group delay handling of the effect chain
was implemented. Based on the Pitch shift effect processing,
using the RubberBand library, the effect produces some amount of latency.
Based on that, if we would like to play the original unprocessed signal
and the processed one together using the wet/dry or wet+dry mode, the two
audio signals will not overlap based on the latency. With that, the common
audio processing approach is to delay the original signal by the amount
of latency to overlap the signals. Based on the Mixxx effect chain API

> (EngineEffectChain), the group delay latency handling was implemented

Reviewer 2: ```suggestion
(`EngineEffectChain`), the group delay latency handling was implemented
```
put class names in backquotes so their appear in monospace.

model implementation, the RubberBand API requires the amount of input samples
and this amount of samples is passed. The main difference between the usage
of these two implementations in the Mixxx app is, which little tradeoff
has to be done. When the “push model” implementation is used, the input
data samples have not to be prefilled between the processing
but the audio dropouts can occur. On the other hand,
the “pull model” implementation with correctly set structure sizes
avoids dropouts, but the input data samples have to be prefilled
before the processing, and a delay will be produced before the output
is produced.  To compare the differences in group delay between the “push model”
and “pull model” implementation, a measurement was made.
The results are captured in the following two graphs.

Based on the first graph, the “pull model” implementation with the size
of the circular buffer of 4096 frames produces a lower latency than
the “push model” even with the latency produced before the output data.
Simultaneously, the delay between dynamic Pitch changes is in a smaller range.

*Push and pull model group delay comparison*

> ![TODO]({static}/images/news/pitch-shift-effect-graph-1.png)

Reviewer 2: would it be possible to remake these with a color scheme that matches the mixxx website better?
Also it would be nice if these were SVGs if possible.
IMO it would also be better if the y-axis would start at 0 instead of the lowest latency value. That would illustrate why either model is imperfect.

| BM_MemCpy/64                        | 4.68 ns | 4.68 ns | 174373476  |
| BM_MemCpy/512                       | 33.0 ns | 33.0 ns | 21221060   |
| BM_MemCpy/4096                      | 154 ns  | 154 ns  | 4535488    |
| BM_StdCpy/64                        | 4.68 ns | 4.68 ns | 152600697  |
| BM_StdCpy/512                       | 32.1 ns | 32.1 ns | 21839231   |
| BM_StdCpy/4096                      | 158 ns  | 158 ns  | 4418188    |
| BM_SampleUtilCopy/64                | 4.68 ns | 4.68 ns | 174349198  |
| BM_SampleUtilCopy/512               | 33.0 ns | 33.0 ns | 21194709   |
| BM_SampleUtilCopy/4096              | 154 ns  | 154 ns  | 4515069    |
| BM_Copy2WithGain/64                 | 8.03 ns | 8.03 ns | 87167131   |
| BM_Copy2WithGain/512                | 64.2 ns | 64.2 ns | 10889704   |
| BM_Copy2WithGain/4096               | 823 ns  | 823 ns  | 847146     |
| BM_Copy2WithRampingGain/64          | 15.6 ns | 15.6 ns | 44785697   |
| BM_Copy2WithRampingGain/512         | 116 ns  | 116 ns  | 6011618    |
| BM_Copy2WithRampingGain/4096        | 1113 ns | 1113 ns | 629003     |

#### Demo video
TODO(davidchocholaty) add demo video

> @Video(https://www.youtube.com/watch?v=rt5Ed5GZ1U8)

Reviewer 2: :stuck_out_tongue_winking_eye: :joy: 
```suggestion
@Video(https://www.youtube.com/watch?v=dQw4w9WgXcQ)
```

Author: :joy::joy::joy:

audio quality and with the same amount of delay. Based on all
the discussions, the RubberBand library has shown as the best option
for the effect, despite the issues which are associated with it. After that,
the Mixxx application did not have implemented the effects delay handler
for the effect chain, so, the implementation of this structure
was automatically needed. Based on that, the goals of the project were changed.
The original proposal contained the Pitch shift effect, and with that
as a project extension to the project requirements of the Mixxx organization,
the Auto-tune effect was proposed. After consideration, the Zulip chat survey
for other Mixxx developers and users was created to be able to vote for possible
project extensions. Based on the survey results, the project extension goal
was changed to the implementation and optimization of the group delay handling
for the effect chain to improve the performance of the Pitch Shift effect
for the wet/dry and wet+dry mode. As the last challenge, I would like to mention
the usage of the std::span from the standard library which is supported
by C++20. Because the Mixxx organization adheres to the Minimum requirements

> policy for the Ubuntu LTS, the EngineEffectsDelay and RingDelayBuffer

Reviewer 2: this is where I'd link to the github wiki page. Also a slight correction: ubuntu/canonical does not really have any minimum requirements policy as far as I'm aware of. They just include whats stable in their repositories. It's our own policy that we want mixxx to be compatible with the latest ubuntu LTS and thus ensure we stay compatible with the dependencies available in ubuntu.

using the “pull model” implementation after the end of the GSoC period.
With that, the wet/dry mode will be done for the effect too.
As the next project extensions, based on the survey, the following options
or features can be added to the Pitch shift effect implementation:

<ul>
    <li>Auto-tune effect</li>
    <li>A piano keyboard interface</li>
    <li>Optimize interface for common controllers</li>
    <li>CPU load balancing</li>
    <li>Consider interaction with the main Pitch shifter</li>
    <li>Expose compensation delay as additional parameter for making
    funny things without extra CPU cycles</li>
</ul>

More widely, as the future work for the Mixxx application, the wider support
for the LV2 standard for effects can be implemented or better, the Carla
audio plugin host can be introduced in the Mixxx application.
It will allow users to use their own effects they like directly with the support
of standards such as  LADSPA, DSSI, LV2, VST2, VST3 and so on,
in the application instead of offering only the built-in effects or poor API
for the LV2 standard. After consultation with my mentor, we agreed,
that I will take on this task as a regular Mixxx contributor after the end

> of GSoC.

Reviewer 2: Just to be clear, you're not forced to do anything mixxx related after gsoc. :sweat_smile:   
I was just suggesting that this is probably a very anticipated feature for mixxx and thus worth spending time on.

Author: Yeah, I know. :smile:  I would like to continue contributing after GSoC and I took it as an agreement, that I can work on this task due to it being from a similar field as the project.

being part of the Mixxx organization and contributing to open source.

#### Conclusion
The wished new effect was implemented, and the issue which requested
this new feature was closed with the "Fix Committed" status.  All requirements
by the Mixxx organization in their project idea, on which the project proposal
was based, were met. Thanks to enough time in the GSoC Coding period was worked
on the project extension. Based on the situation and the importance of new
Mixxx features, the originally proposed extension was replanned and changed.
The new group delay handling structure was successfully implemented
and optimized with the implementation of the extended data structure.
In the GSoC Coding period, the effect was started to optimized and polished
for minimizing the latency.

During the coding period, I lined up among the top Mixxx contributors
for the last month, with 61 commits authored, and I became
the 27th of 238 contributors for the Mixxx application, with 119 commits
in total.

*August 2022 mixxxdj/mixxx contributors*
![TODO]({static}/images/news/august-2022-contributors.png)

*Contributor summary*

> ![TODO]({static}/images/news/contributor-summary.png)

Reviewer 2: Can you redo these screenshots with a dark theme to fit the website better? I think you can find that on github > settings > appearance.

#### Acknowledgements

> First, I would like to many thank my mentor Swiftb0y for his guidance, help,
> reviews, and a lot of new information and lessons he gave me during the summer.

Reviewer 2: Thank you. I'd also like to thank you for spending your valuable time with me on mixxx. While certainly stressful, it has still been fun with you.

You can link my author page if you want.
```suggestion
First, I would like to many thank my mentor [@Swiftb0y]({author}nikolaus-einhauser) for his guidance, help,
reviews, and a lot of new information and lessons he gave me during the summer.
```

Author: Thank you very much. Yeah, it was fun for me too. :+1: 

Perfect, I will add it. Is okay that I used your nickname? Or would you like to replace it with your full name as on the developer link? I used the names you and Daniel used on the GSoC page as Mentor's names, so I followed them.

Reviewer 2: I don't really care, my nickname and my real name have become synonymous. Use whatever you like.

title: GSoC 2022 Work Product - Pitch Shift effect and Group delay handling
authors: David Chocholatý
status: draft

> tags: TODO(davidchocholaty) add tags

Reviewer 2: ```suggestion
tags: gsoc, gsoc-2022
```
Any other tags that would fit?

Author: If I am thinking about it, just the latency offers, but not sure.

#### Pull requests and issues
[mixxx#4775](https://github.com/mixxxdj/mixxx/pull/4775)
- PitchShiftEffect: add independent effect

The PR adds an independent effect to Mixxx's built-in effects.
The implementation uses the RubberBand library for changing a pitch
of an input track. The effect works in real-time mode and adheres
to the “push model” implementation. It means that the input data are offered
to the RubberBand library instead of that the library requires the amount
of input data.

---

[mixxx#4810](https://github.com/mixxxdj/mixxx/pull/4810)
- EngineEffectsDelay: effect chain delay handling

This PR adds the structure for the group delay handling of the effect chain.
Based on that, some effects can produce latency due to their inner processing.
The latency has to be handled for the wet/dry and wet+dry modes that the dry
and wet signals overlapped. The structure for delay reporting from the effects
into the effect chain was implemented. With that, the dry signal delaying
to overlap with the wet signal was implemented too. Because it is
a critical part of the application engine performance, the tests and benchmarks
were included in the development.

> In this PR, the std::span was newly introduced into the Mixxx app code

Reviewer 2: ```suggestion
In this PR, the `std::span` was newly introduced into the Mixxx app code
```

---

[mixxx#4852](https://github.com/mixxxdj/mixxx/pull/4852)
- RingDelayBuffer: ring buffer for delay handling

During the creation of the `EngineEffectsDelay` for the group delay handling
of the effect chain, it was suggested to create an optimized data structure
for the inner processing based on the ring buffer. This widely-known
signal processing structure was improved and optimized specifically
for the use case with handling of delay. Again, tests and benchmarks
were created for the `RingDelayBuffer` and based on benchmarks the used data
copy functions were compared.

---

[vcpkg#48](https://github.com/mixxxdj/vcpkg/pull/48)
- [rubberband] add overlaid rubberband v3

During the coding period, the new RubberBand library release v3.0.0
was announced. Based on the implementation for adding RubberBand v2.0.2 directly
into the microsoft / vcpkg repository by the Mixxx organization admin,
the RubberBand v3.0.0 was added into the overlaid ports in the Mixxx fork
of the original repository.

---

[mixxx#4869](https://github.com/mixxxdj/mixxx/pull/4869)
- EngineFilterDelay: clamp wrong delay values

Based on the [mixxx#4810](https://github.com/mixxxdj/mixxx/pull/4810) implementation

> it was figured out the bug in the `EngineFilterDelay` structure was. The structure

Reviewer 2: ```suggestion
While working on [mixxx#4810](https://github.com/mixxxdj/mixxx/pull/4810) I encountered a bug in the `EngineFilterDelay` structure: The structure
```

works in a similar way but for a little different use case.
Newly the unacceptably huge delay values are clamped in the setter, so,
based on the inner calculation the structure will not produce absolutely wrong
output. The PR was merged the same day as its creation.

---

[mixxx#4898](https://github.com/mixxxdj/mixxx/pull/4898)
- PitchShiftEffect: decrease and report latency

In this draft PR was worked as another project extension. The implemented “push”
way model is extended into the “pull” model instead. The new approach decreases
the effect latency and this latency is reported in the effect chain
delay handler. This PR is still a Work In Progress. As the last work done
the latency measurements were performed for several implementations
and for different pitch settings. The measured data was plotted
for demonstration. The new implementation was accepted and the PR will be done
in the non-GSoC time as a future Mixxx contributor. With the new implementation,
the Mixxx circular buffer data structure was improved and optimized
for performance. So, it remains to finish the pull implementation by setting
the right size of the input ring buffer.  Eventually, implement the input

> ring buffer size depending on the range, that was set. As the last thing,

Reviewer 2: ```suggestion
ring buffer size depending on the range that was set. As the last thing,
```

the valid delay value propagation for the effect will be finished.

---

[mixxx#4901](https://github.com/mixxxdj/mixxx/pull/4901)
- PitchShiftEffect: extend effect options

The PR extends options of the Pitch shift effect. The Range knob is added
to the setting of the range of Pitch knob. These two knobs work similarly
to the real professional Pioneer DJM-900NX2 mixer which is widely used
in clubs for live DJ mixing. With that, the Semitones mode toggle was added
for changing the scale of the Pitch knob.  By default, this toggle is on,
and the Pitch knob works in the semitones mode. In musical terminology,
the pitch is changed based on the semitone chromatic scale. If the toggle
is off, the Pitch knob works in the continuous mode, as is the default

> for the RubberBand library. As last, the Formant preserving option was added

Reviewer 2: ```suggestion
is off, the Pitch knob works in the continuous mode, which is also the default
in the RubberBand library. At last, the Formant preserving option was added
```

- RingDelayBufferTest: refactor includes and span creation

The tests for the `RingDelayBuffer` are refactored and the span creations
are deduplicated.

---

[mixxx#10858](https://github.com/mixxxdj/mixxx/pull/10858)
- PitchShiftEffect: add description comments

Added the comments for the Pitch shift effect processing.

---

[website#279](https://github.com/mixxxdj/website/pull/279)
- content/news: add GSoC 2022 Work Product

Adds a blog post containing the *"Work Product"* for Google Summer of Code 2022
on the Mixxx website.

---

#### Implementation

##### *Pitch shift effect*
The Pitch shift effect main algorithm is implemented
in the `PitchShiftEffect` class. The implementation uses the widely known
audio time-stretching and pitch-shifting library RubberBand.
The implementation adheres to the “push model”. That means, that
the input audio samples are offered to the RubberBand library API directly.
Instead of the main Pitch shifter for each deck player, which
has a limited range (only 7 semitones up and down in musical terms,
which means not even an octave), the independent Pitch shift effect offers
to work in the range of + 2 octaves (+ 24 semitones) and - 2 octaves

> (-24 semitones). The pitch shift effect has the following options:

Reviewer 2: ```suggestion
to work in the range of ± 2 octaves (± 24 semitones). 
The pitch shift effect has the following options:
```

and “pull model” implementation, a measurement was made.
The results are captured in the following two graphs.

Based on the first graph, the “pull model” implementation with the size
of the circular buffer of 4096 frames produces a lower latency than
the “push model” even with the latency produced before the output data.
Simultaneously, the delay between dynamic Pitch changes is in a smaller range.

*Push and pull model group delay comparison*
![TODO]({static}/images/news/pitch-shift-effect-graph-1.svg)

In the second graph, the group delay for the unchanged pitch was measured.
The differences in the measurements are, that the previous pitch values
before the unchanged pitch settings are different. With the results
of the last measurement, the “pull model” implementation is more stable
and produces a similar, almost constant delay after dynamic pitch changes
for the same pitch setting.

*Push and pull model group delay comparison for unchanged pitch*

![TODO]({static}/images/news/pitch-shift-effect-graph-2.svg)

The results of the measurements clearly show, that despite the “pull model”
implementation for the Pitch shift effect in the effect chain is not optimal,

> it should be preferred before the “push model” implementation.

Reviewer 2: ```suggestion
it should be preferred over the “push model” implementation.
```

| BM_Copy2WithGain/64                 | 8.03 ns | 8.03 ns | 87167131   |
| BM_Copy2WithGain/512                | 64.2 ns | 64.2 ns | 10889704   |
| BM_Copy2WithGain/4096               | 823 ns  | 823 ns  | 847146     |
| BM_Copy2WithRampingGain/64          | 15.6 ns | 15.6 ns | 44785697   |
| BM_Copy2WithRampingGain/512         | 116 ns  | 116 ns  | 6011618    |
| BM_Copy2WithRampingGain/4096        | 1113 ns | 1113 ns | 629003     |

#### Demo video
TODO(davidchocholaty) add demo video

@Video(https://www.youtube.com/watch?v=rt5Ed5GZ1U8)

#### Challenges
Several problems arose during the coding period.
Despite the medium project size, it was needed to spend much more time
working on the project based on the issues. I think the biggest challenge
in this project was exactly the implementation of the Pitch Shift effect
using the RubberBand library. The issue occurred soon, that the effect chain
offers the effect to work only with the fixed size audio chunks. Based on that,
it is not possible to require an amount of input audio data as needed
for the RubberBand library “pull model” implementation. The SoundTouch library

> for Pitch Shifting was tested too but produced results with the worse

Reviewer 2: ```suggestion
for Pitch Shifting was tested too but produced results with worse
```

by C++20. Because the Mixxx organization adheres to the own
[Minimum requirements policy](
 https://github.com/mixxxdj/mixxx/wiki/Minimum%20requirements%20policy)
for the Ubuntu LTS, the `EngineEffectsDelay` and `RingDelayBuffer`
could be merged after the official Ubuntu release was announced in the middle
of August due to support of C++20.

#### Future work
Concretely for the Pitch Shift effect, the effect will be improved
using the “pull model” implementation after the end of the GSoC period.
With that, the wet/dry mode will be done for the effect too.
As the next project extensions, based on the survey, the following options
or features can be added to the Pitch shift effect implementation:

* Auto-tune effect
* A piano keyboard interface
* Optimize interface for common controllers
* CPU load balancing
* Consider interaction with the main Pitch shifter
* Expose compensation delay as additional parameter for making
 funny things without extra CPU cycles

More widely, as the future work for the Mixxx application, the wider support
for the LV2 standard for effects can be implemented or better, the Carla
audio plugin host can be introduced in the Mixxx application.

> It will allow users to use their own effects they like directly with the support
> of standards such as  LADSPA, DSSI, LV2, VST2, VST3 and so on,

Reviewer 2: ```suggestion
It will allow users to use their favorite effects enabled via audio plugin
standards such as  LADSPA, DSSI, LV2, VST2, VST3 and so on,
```

that it was the best experience so far for me as a developer I ever had.
It is awesome, that I can publish my work and have immediately the feedback
and proposed improvements. With that, I liked the open source development
to just how much I can learn by the awesome people and create new cool stuff.
Despite I’m a college student, this actually missed me a lot,
to just have feedback on my work which opens me the opportunities to learn.
I really felt that I’m a part of the community. I will be happy to continue
being part of the Mixxx organization and contributing to open source.

#### Conclusion
The wished new effect was implemented, and the issue which requested
this new feature was closed with the "Fix Committed" status.  All requirements
by the Mixxx organization in their project idea, on which the project proposal
was based, were met. Thanks to enough time in the GSoC Coding period was worked
on the project extension. Based on the situation and the importance of new
Mixxx features, the originally proposed extension was replanned and changed.
The new group delay handling structure was successfully implemented
and optimized with the implementation of the extended data structure.

> In the GSoC Coding period, the effect was started to optimized and polished
> for minimizing the latency.

Reviewer 2: ```suggestion
In the GSoC Coding period, work was started to minimize the effects latency as
well as to polish the effect even more. Unfortunately that work could 
not be finished before the GSoC deadline. 
```

Author: Sounds cool, really well-formed. :+1:

During the coding period, I lined up among the top Mixxx contributors
for the last month, with 61 commits authored, and I became
the 27th of 238 contributors for the Mixxx application, with 119 commits
in total.

*August 2022 mixxxdj/mixxx contributors*
![TODO]({static}/images/news/august-2022-contributors.png)

*Contributor summary*
![TODO]({static}/images/news/contributor-summary.png)

#### Acknowledgements
First, I would like to many thank my mentor Swiftb0y for his guidance, help,
reviews, and a lot of new information and lessons he gave me during the summer.

> I'm just motivated and learned a lot. I would like thank to Mixxx organization

Reviewer 2: ```suggestion
I'm just motivated and learned a lot. I would like to thank the Mixxx organization
```

| BM_WriteReadWholeBufferDelay/4096   | 1500 ns | 1528 ns | 457952     |
| BM_MemCpy/64                        | 4.68 ns | 4.68 ns | 174373476  |
| BM_MemCpy/512                       | 33.0 ns | 33.0 ns | 21221060   |
| BM_MemCpy/4096                      | 154 ns  | 154 ns  | 4535488    |
| BM_StdCpy/64                        | 4.68 ns | 4.68 ns | 152600697  |
| BM_StdCpy/512                       | 32.1 ns | 32.1 ns | 21839231   |
| BM_StdCpy/4096                      | 158 ns  | 158 ns  | 4418188    |
| BM_SampleUtilCopy/64                | 4.68 ns | 4.68 ns | 174349198  |
| BM_SampleUtilCopy/512               | 33.0 ns | 33.0 ns | 21194709   |
| BM_SampleUtilCopy/4096              | 154 ns  | 154 ns  | 4515069    |
| BM_Copy2WithGain/64                 | 8.03 ns | 8.03 ns | 87167131   |
| BM_Copy2WithGain/512                | 64.2 ns | 64.2 ns | 10889704   |
| BM_Copy2WithGain/4096               | 823 ns  | 823 ns  | 847146     |
| BM_Copy2WithRampingGain/64          | 15.6 ns | 15.6 ns | 44785697   |
| BM_Copy2WithRampingGain/512         | 116 ns  | 116 ns  | 6011618    |
| BM_Copy2WithRampingGain/4096        | 1113 ns | 1113 ns | 629003     |

> 
> #### Demo video
> TODO(davidchocholaty) add demo video
> 
> @Video(https://www.youtube.com/watch?v=rt5Ed5GZ1U8)

Reviewer 2: Lets hide this for now until we have the correct video uploaded.
```suggestion

<!---
#### Demo video
TODO(davidchocholaty) add demo video

@Video(https://www.youtube.com/watch?v=rt5Ed5GZ1U8)
-->
```

     that usually come from a direct recording. On the other hand,
     wet sounds refer to the processed sound/signal[^8].
* Ring buffer (circular buffer)
    * In computer science, a circular buffer or ring buffer
     is a data structure that uses a single, fixed-size buffer
     as if it were connected end-to-end. This structure lends itself
     easily to buffering data streams[^9].
* Audio buffer
    * An audio buffer holds a fixed size amount of sampled audio data.
     The audio buffer size determines the time allowed for the computer
     to process the audio data. Thus, it also determines Latency.
* Latency
    * In the audio world, “latency” is another word for “delay”[^10].
     The latency (time) of an audio system refers to the time difference
     from the moment a signal is fed into the system, to the moment it appears
     at the output[^11]. For example, audio latency is when there’s a noticeable
     delay between the sound being played and the moment it reaches
     the speakers[^10]. Depending on the application, such a delay can have various
     effects. Usually, the aim is to achieve the lowest possible latency[^11].

> 

Reviewer 2: `10` and `11` are here twice, is that on purpose?

Author: Yes, it was meant to be. The reason for that is, that the description is a combination of citations from two sources and the whole paragraph made more sense to me if the sentences have combined these ways, so the two citation sources alternate.

and the `RingDelayBuffer` as an optimized data structure for the same use case,
the tests were included with the use of the GoogleTest framework. Basically,
the common situations were tested, then extreme cases and cases
that are not allowed but have to be handled for the release builds.
On the basis that these implemented structures are critical
from the point of view of performance, with tests, the benchmarks were created
with the use of the Google Benchmark. Based on the results of benchmarks,
the used functions and algorithms were compared. After both of the mentioned
structures were implemented, tested and optimized, the `RingDelayBuffer`
was introduced in the `EngineEffectsDelay` as an inner structure
for the group delay handling. Based on the changes and use of optimized
data structure the performance has highly improved. The performance differences
are shown in the following benchmarks results taken over
from the Ubuntu GitHub CI’s results.

##### *Without RingDelayBuffer*

Run on (2 X 2593.91 MHz CPU s)  
CPU Caches:

* L1 Data 32 KiB (x2)
* L1 Instruction 32 KiB (x2)
* L2 Unified 1024 KiB (x2)
* L3 Unified 36608 KiB (x1)

> Load Average: 1.39, 1.82, 1.75

Reviewer 2: What does this signify? Load on what? What are those numbers?

Author: To be honest, I had to do some research about the meaning :smile:. In a nutshell, it describes the computational work of the CPU. More information in the following sources:

https://en.wikipedia.org/wiki/Load_(computing)
https://www.guyrutenberg.com/2008/07/25/understanding-load-average-a-practitioner-guide/

IMO because the pattern is written exactly the same on the Linux system (the `uptime` command), I propose to just keep it as it is for the user who is consecration in it (others will just probably skip this part) or remove it if we agree that it doesn't make sense to publish. Because it is exactly the copy-paste from the Google Benchmark output, I would prefer to keep it as is. IMHO I don't see much value in explaining the reason for the resulting output.

What's your opinion?

Reviewer 2: Interesting, I didn't know google benchmark outputted this info. To be honest, I'd remove them because they just signify load averages over the last 1, 5, 15 minutes. For a benchmark that only runs for 30s, I don't see much value in the info retrieved with the mentioned window sizes.

Author: Sounds reasonable, I agree.

| BM_ZeroDelay/512                   | 2079 ns  | 2078 ns  | 336963     |
| BM_ZeroDelay/4096                  | 16616 ns | 16614 ns | 42139      |
| BM_DelaySmallerThanBufferSize/64   | 269 ns   | 269 ns   | 2600146    |
| BM_DelaySmallerThanBufferSize/512  | 2119 ns  | 2119 ns  | 330263     |
| BM_DelaySmallerThanBufferSize/4096 | 16978 ns | 16977 ns | 40822      |
| BM_DelayGreaterThanBufferSize/64   | 269 ns   | 269 ns   | 2597728    |
| BM_DelayGreaterThanBufferSize/512  | 2125 ns  | 2125 ns  | 329431     |
| BM_DelayGreaterThanBufferSize/4096 | 16979 ns | 16978 ns | 40976      |
| BM_DelayCrossfading/64             | 608 ns   | 608 ns   | 1151956    |
| BM_DelayCrossfading/512            | 4865 ns  | 4865 ns  | 143875     |
| BM_DelayCrossfading/4096           | 40081 ns | 40079 ns | 17621      |
| BM_DelayNoCrossfading/64           | 539 ns   | 539 ns   | 1298494    |
| BM_DelayNoCrossfading/512          | 4238 ns  | 4238 ns  | 165164     |
| BM_DelayNoCrossfading/4096         | 33939 ns | 33936 ns | 20622      |

##### *With RingDelayBuffer*

Run on (2 X 2593.91 MHz CPU s)  
CPU Caches:

* L1 Data 32 KiB (x2)
* L1 Instruction 32 KiB (x2)
* L2 Unified 1024 KiB (x2)

> * L3 Unified 36608 KiB (x1)

Reviewer 2: ```suggestion
Run on the same system as above.
```

Author: Thank you, great idea. :+1:

| BM_MemCpy/512                       | 33.0 ns | 33.0 ns | 21221060   |
| BM_MemCpy/4096                      | 154 ns  | 154 ns  | 4535488    |
| BM_StdCpy/64                        | 4.68 ns | 4.68 ns | 152600697  |
| BM_StdCpy/512                       | 32.1 ns | 32.1 ns | 21839231   |
| BM_StdCpy/4096                      | 158 ns  | 158 ns  | 4418188    |
| BM_SampleUtilCopy/64                | 4.68 ns | 4.68 ns | 174349198  |
| BM_SampleUtilCopy/512               | 33.0 ns | 33.0 ns | 21194709   |
| BM_SampleUtilCopy/4096              | 154 ns  | 154 ns  | 4515069    |
| BM_Copy2WithGain/64                 | 8.03 ns | 8.03 ns | 87167131   |
| BM_Copy2WithGain/512                | 64.2 ns | 64.2 ns | 10889704   |
| BM_Copy2WithGain/4096               | 823 ns  | 823 ns  | 847146     |
| BM_Copy2WithRampingGain/64          | 15.6 ns | 15.6 ns | 44785697   |
| BM_Copy2WithRampingGain/512         | 116 ns  | 116 ns  | 6011618    |
| BM_Copy2WithRampingGain/4096        | 1113 ns | 1113 ns | 629003     |

<!---
#### Demo video
TODO(davidchocholaty) add demo video

@Video(https://www.youtube.com/watch?v=rt5Ed5GZ1U8)

> -->

Reviewer 3: @Be-ing uploaded the video to our YouTube channel: https://www.youtube.com/watch?v=NTjV7s5Jb_o

Author: Awesome. :+1:  Thank you very much for the information and @Be-ing for so fast uploading.

the usage of the std::span from the standard library which is supported
by C++20. Because the Mixxx organization adheres to the own
[Minimum requirements policy](
 https://github.com/mixxxdj/mixxx/wiki/Minimum%20requirements%20policy)
for the Ubuntu LTS, the `EngineEffectsDelay` and `RingDelayBuffer`
could be merged after the official Ubuntu release was announced in the middle
of August due to support of C++20.

#### Future work
Concretely for the Pitch Shift effect, the effect will be improved
using the “pull model” implementation after the end of the GSoC period.
With that, the wet/dry mode will be done for the effect too.
As the next project extensions, based on the survey, the following options
or features can be added to the Pitch shift effect implementation:

* Auto-tune effect
* A piano keyboard interface
* Optimize interface for common controllers
* CPU load balancing
* Consider interaction with the main Pitch shifter
* Expose compensation delay as additional parameter for making
 funny things without extra CPU cycles

More widely, as the future work for the Mixxx application, the wider support

> for the LV2 standard for effects can be implemented or better, the Carla[^2]

Reviewer 2: This footnote seems wrong.

Author: Yeah, right. Thank you! I forgot to remove it.

[mixxx#10858](https://github.com/mixxxdj/mixxx/pull/10858)
- PitchShiftEffect: add description comments

*Status: Merged*

Added the comments for the Pitch Shift effect processing.

---

[website#279](https://github.com/mixxxdj/website/pull/279)
- content/news: add GSoC 2022 Work Product

*Status: Merged*

Adds a blog post containing the *"Work Product"* for Google Summer of Code 2022
on the Mixxx website.

---

#### Implementation

##### *Pitch Shift effect*
The Pitch Shift effect main algorithm is implemented
in the `PitchShiftEffect` class. The implementation uses the widely known
audio time-stretching and pitch-shifting library RubberBand.
The implementation adheres to the “push model”. That means, that
the input audio samples are offered to the RubberBand library API directly.
Instead of the main Pitch Shifter for each deck player, which
has a limited range (only 7 semitones up and down in musical terms,
which means not even an octave), the independent Pitch Shift effect offers
to work in the range of ± 2 octaves (± 24 semitones).
The Pitch Shift effect has the following options:

* Pitch knob
* Range knob
* Semitones mode
* Formant preserving

> The Pitch knob changes a pitch of a track up or down. For the default

Reviewer 2: ```suggestion
The Pitch knob changes the pitch of a track up or down. For the default
```

middle position, the track pitch is unchanged. The Range knob ensures setting

> a range of the Pitch knob. The Pitch knob based on the range setting can work

Reviewer 2: ```suggestion
the range of the Pitch knob. The Pitch knob based on the range setting can work
```

process and for constructive criticism which offers me learn many new things
in the past weeks. I would like to thank my summer colleague for the Mixxx
organization and friend [@Fatih Emre](https://github.com/fatihemreyildiz)
for his help and synergy on the final blog structure and chapters. Thank you

> to [@Be.](http://127.0.0.1:8000/news/author/be) for taking care of publishing

Reviewer 2: ```suggestion
to [@Be.]({author}be) for taking care of publishing
```

Reviewer 4: You made the video. Uploading it was trivial. You don't need to credit me for that.

Author: Right :smile:

Author: Sorry for the possible misunderstanding. The previous deleted ("Right :smile:") answer was for a proposal for change by @Swiftb0y. My web browser just did not update the page. @Be-ing I'm grateful to you that you publish the video so early and made adding it to this PR possible. Of course, I understand if you don't want to be listed, it's your choice. Please for your feedback.

Reviewer 4: I'd prefer to not be credited here.

Author: Okay, no problem. I will fix it.
